lib/modifiers/flatten.js

```diff
@@ -0,0 +1,60 @@
+var _ = require('lodash');
+var through = require('through2');
+
+/**
+ * Returns a new stream that flattens all arrays passing through by one level. All non-array
+ * items will be passed through as-is.
+ *
+ * @method
+ * @static
+ * @since    1.5.0
+ * @category Util
+ *
+ * @returns {Stream} - Transform stream.
+ *
+ * @example
+ * var input = through.obj();
+ *
+ * input.pipe(nodeStream.flatten());
+ *
+ * input.write([1, 2, 3]);
+ * input.write([4, 5]);
+ * input.write(6);
+ *
+ * // => [1, 2, 3, 4, 5, 6]
+ * @example
+ * var input = through.obj();
+ *
+ * input
+ *   // batch items that are read
+ *   .pipe(nodeStream.batch({ count: 2 }))
+ *   // perform a transform action on the batches
+ *   .pipe(transformBatch())
+ *   // flatten the batches back
+ *   .pipe(nodeStream.flatten());
+ *
+ * input.write(1);
+ * input.write(2);
+ * input.write(3);
+ * input.write(4);
+ * input.write(5);
+ *
+ * // => [1, 2, 3, 4, 5]
+ */
+function flatten() {
+  return through.obj(function Transform(data, enc, cb) {
+    var self = this;
+
+    if (!_.isArray(data)) {
+      return cb(null, data);
+    }
+
+    data.forEach(function (val) {
+      self.push(val);
```

Comment:
I actually did think about this. So, your assumption is actually not true. Write/reads are synchronous (meaning, you write and the consumer may immediately read synchronously... just like an event emitter). This is the output reflecting that (flattening `_.times(10)` with an on `data` attached):

```
pushing 0
read 0
pushing 1
read 1
pushing 2
read 2
pushing 3
read 3
pushing 4
read 4
pushing 5
read 5
pushing 6
read 6
pushing 7
read 7
pushing 8
read 8
pushing 9
read 9
```

That's not to say there isn't any issues, necessarily. This implementation doesn't really handle backpressure well. It is possibly that the stream gets paused half-way through writing a single array to it (if it gets paused between, I believe `through` and the pipe implementation would just pause the stream we are reading from). I feel like I need to think about this a bit more.